### PR TITLE
Add an --all-cores command line option to scons

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixup for #16868

### Summary of the issue:
As part of #16868, i intended to implement support to pass a number of 0 cores to scons (e.g. `scons source -j0`) to automatically pick all available cores. It turns out that you cant override an option with `SetOption` when set on the command line.

### Description of user facing changes
None, build system related.

### Description of development approach
Added an `--all-cores` parameter that will pick all available cores.

### Testing strategy:
Ensured that the build was marvelously faster when using the `--all-cores` parameter.

### Known issues with pull request:
None known

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
	- Introduced a command-line option `--all-cores` for the build process, allowing users to utilize all available CPU cores for improved performance.
	- Updated documentation to guide users on using the `--all-cores` option in the `scons` build tool.

- **Bug Fixes**
	- Improved documentation clarity by changing command parameters from `-j0` to `--all-cores`. 

- **Documentation**
	- Enhanced user guidance on optimizing build performance with the new `--all-cores` parameter.

### CODE REVIEW

**Comment by dpy013** [comment]
This feature is really good, but it doesn't showcase its performance on AppVeyor. Will it be faster using GitHub Actions?

**Comment by LeonarddeR** [comment]
The primary reason why this is not used on appveyor is because it scrambles the output.

**Comment by SaschaCowley** [comment]
Note that you can set this to be the case always by adding `-j N` (where n is the number of cores you want to use) to the `SCONSFLAGS` environment variable. Providing `-j N` at the CLI will still override it for cases when you need to inspect the output. So for example, I have

```ps
PS D:\projects\nvda> echo $env:SCONSFLAGS
-j 16
```

File: sconstruct [hunk 1]
```diff
@@ -159,19 +159,28 @@ env = Environment(
 # speed up subsequent runs by checking timestamps of targets and dependencies, and only using md5 if timestamps differ.
 env.Decider("MD5-timestamp")
 
+AddOption(
+	"--all-cores",
+	action="store_true",
+	dest="all_cores",
+	default=False,
+	help="always use the maximum number of available processor cores",
```
Comment (SaschaCowley) [inline sconstruct:167]: This implies the option is persisted across sessions.
```suggestion
	help="Use the maximum number of available processor cores",
```

File: sconstruct [hunk 2]
```diff
@@ -159,19 +159,28 @@ env = Environment(
...
+if (numJobs := GetOption("num_jobs")) < numCores:
+	msg = "Warning: "
+	if allCores:
+		msg += "Providing both the '--all-cores' and '-j'/'--num-jobs' parameters is not supported. "
+	msg += (
+		f"Building with {numJobs} concurrent job{'s' if numJobs != 1 else ''} "
 		f"while {numCores} CPU threads are available. "
 		f"Running SCONS with the parameter '-j{numCores}' may lead to a faster build. "
-		"Running SCONS with parameter '-j0' will automatically pick all available cores. "
+		"Running SCONS with parameter '--all-cores' will automatically pick all available cores."
```
Comment (seanbudd) [inline sconstruct:181]:
```suggestion
		"Running SCONS with parameter '--all-cores' will automatically pick all available cores. "
```

**Review by seanbudd — APPROVED** [review]

**Review by SaschaCowley — CHANGES_REQUESTED** [review]
Thanks for fixing this.

**Review by SaschaCowley — APPROVED** [review]

**Pull request merged**